Dockerfile

```diff
@@ -44,9 +44,9 @@ RUN dnf install -y python3 python3-requests && \
     dnf upgrade -y && \
     dnf --setopt=install_weak_deps=False install -y $(cat /tmp/main-packages-list.txt) && \
     dnf clean all && \
-    rm -rf /var/cache/{yum,dnf}/*
+    rm -rf /var/cache/{yum,dnf}/* && \
+    mkdir -p /tftpboot
```

Comment:
Maybe I'm missing something, but it doesn't seem like merging these steps into 1 command is really the best approach. Some of the layers aren't going to be different from build to build, so could be cached, right? Like if this mkdir command came first in its own RUN command, that layer of the image would almost never change.

Having all of the dnf commands together in one step does make sense. Is the curl command setting something up for dnf to be able to install the tripleo packages, or is that independent?

Reply:
Please consider that when processing a RUN command the files updated in the container are not examined to determine if a cache hit exists. In that case just the command string itself is used to find a match.
In this specific case, having one RUN command is better than the two RUN commands we had before, I'm not sure having the mkdir first, in a different RUN command, would change much, but we can try.
The layer created using RUN with all the dnf commands won't be cached, so it doesn't make sense to decouple them.
Also consider that having dnf alone in a RUN line could cause caching issues.
The curl command enables the tripleo repos and it's necessary to install all the openstack packages, I would leave it there.

Reply:
The unchanging layers wouldn't be cached during the build, but would be cached when pushing or pulling them, right? In this case the layer with the directory in it isn't very big, so it would hardly matter. But as a general approach it seems like lumping all changes into a single layer makes it impossible to take advantage of any caching at all.

Reply:
The layer cache is just for the building process.
In general best practice says to merge all the RUN commands to avoid creating multiple layers, because these rarely invalidate the cache.
The RUN command will invalidate the cache only if the command string change, so this change is useful to reduce the number of layers, both during build as during push/pull.
For example, using pull, by default three layers of an image at a time are pulled by the docker daemon, so reducing the number of layers will reduce the pull time.